adapters/ovsdb/Main.hs

```diff
@@ -41,81 +46,93 @@ data TOption = OVSFile     String
-addOption config (OVSFile f) = do
-    when (confOVSFile config /= "") $ errorWithoutStackTrace "Multiple input files specified"
-    return config {confOVSFile = f}
-addOption config (OutputTable t) = return config{ confOutputTables = nub ((t,[]) : confOutputTables config)}
-addOption config (ProxyTable t) = return config{ confProxyTables = nub (t : confProxyTables config)}
-addOption config (KeyColumn c) = do
+addOption :: (Action, Config) -> TOption -> IO (Action, Config)
+addOption (_, config) Version = do return (ActionVersion, config)
+addOption (a, config) (OVSFile f) = do
+    when (ovsFile config /= "") $ errorWithoutStackTrace "Multiple input files specified"
+    return (a, config {ovsFile = f})
+addOption (a, config) (OutputTable t) = return (a, config{ outputTables = nub ((t,[]) : outputTables config)})
+addOption (a, config) (ProxyTable t) = return (a, config{ proxyTables = nub (t : proxyTables config)})
+addOption (a, config) (KeyColumn c) = do
     case splitOn "." c of
          [table, col] -> do
-            when (isNothing $ lookup table $ confOutputTables config)
+            when (isNothing $ lookup table $ outputTables config)
                  $ errorWithoutStackTrace $ "Unknown output table name " ++ table
-            return $ config{confKeys = M.alter (maybe (Just [col]) (\keys -> Just $ nub $ col:keys)) table
-                                               $ confKeys config}
+            return $ (a, config{keys = M.alter (maybe (Just [col]) (\keys -> Just $ nub $ col:keys)) table
+                                               $ keys config})
          _ -> errorWithoutStackTrace $ "Invalid column name " ++ c
-addOption config (ROColumn c) = do
+addOption (a, config) (ROColumn c) = do
     case splitOn "." c of
          [table, col] -> do
-            when (isNothing $ lookup table $ confOutputTables config)
+            when (isNothing $ lookup table $ outputTables config)
                  $ errorWithoutStackTrace $ "Unknown output table name " ++ table
             let outtabs = map (\(t,ro) -> if t == table then (t, nub $ col:ro) else (t,ro))
-                              $ confOutputTables config
-            return $ config{confOutputTables = outtabs}
+                              $ outputTables config
+            return $ (a, config{outputTables = outtabs})
          _ -> errorWithoutStackTrace $ "Invalid column name " ++ c
+addOption (a, _) (ConfigJsonI c) = do
+    cdata <- readFile c
+    return $ (a, fromJust $ ((decode $ LZ.pack cdata) :: Maybe Config))
+addOption (a, config) (ConfigJsonO c) = do
+    writeFile c $ LZ.unpack $ encode config
+    return (a, config)
 
-validateConfig :: Config -> IO ()
-validateConfig Config{..} = do
-    when (confOVSFile == "" && confAction == ActionCompile) $ errorWithoutStackTrace "Input file not specified"
+validateConfig :: (Action, Config) -> IO ()
+validateConfig (action, Config{..}) = do
```

Comment:
This needs to additionally validate that the user does not specify both `-c` and one of the other options that will end up being ignored.

Reply:
This is done in addOption for ConfigJsonI.

Reply:
But that only helps if -c occurs after other options, bit before

Reply:
I think this behavior is very sensible: essentially options are processed from left to right, and you can load a configuration file and override some of its settings.